Adding topic to update Log Forward CR (#26161)
openshift/openshift-docs: modules/cluster-logging-log-forward-update.adoc

// * logging/cluster-logging-external.adoc

// Might not be needed. See https://issues.redhat.com/browse/LOG-654

[id="cluster-logging-log-forward-update_{context}"]
= Updating log forwarding custom resources

The {product-title} Log Forward API has been promoted from Technology Preview to Generally Available in {product-title} 4.6. The GA release contains some improvements and enhancements that require you to make a change to your Cluster Logging custom resource (CR) and to replace your Log Forwarding custom resource (CR) with a `ClusterLogForwarder` instance.

.Procedure 

If you use the Log Forwarding API to forward logs, after the Elasticsearch Operator and Cluster Logging Operator are fully updated to 4.6, you must make the following modifications:

. Edit the Cluster Logging custom resource (CR) to remove the `logforwardingtechpreview` annotation:
+
.Sample Cluster Logging CR
[source,yaml]
----
apiVersion: "logging.openshift.io/v1"
kind: "ClusterLogging"
metadata:
  annotations:
    clusterlogging.openshift.io/logforwardingtechpreview: enabled <1>
  name: "instance"
  namespace: "openshift-logging"
....

----
<1> Remove the remove the `logforwardingtechpreview` annotation.

> . Edit the Log Forwarding CR to create a `ClusterLogForwarder` instance:

Reviewer: We couldn't edit the TP Forwarding CR directly. We should export it, modify locally and create a new GA LF CR
oc get LogForwarding instance -o yaml| tee ClusterLogForwarder.yaml

For example http://pastebin.test.redhat.com/909891

Drop the useless fields  like: .status, metadata.managedFields .creationTimestamp and etc.

   - name: elasticsearch
     type: "elasticsearch"
     url: http://remote.elasticsearch.com:9200 <4>
  pipelines:
   - name: mylogs
      inputRefs: <5>
        - infrastructure
        - application   
        - audit
     outputRefs:
     - elasticsearch
   - name: auditlogs
      inputRefs: <5>
        - audit
     outputRefs:
        - default <6>
----
<1> Change the `apiVersion` from: `"logging.openshift.io/v1alpha1"` to `"logging.openshift.io/v1"`.
<2> Change the object kind from `kind: "LogForwarding"` to `kind: "ClusterLogForwarder"`.
<3> Remove the `disableDefaultForwarding: true` parameter.
<4> Change the output parameter from `spec.outputs.endpoint` to `spec.outputs.url`. Add a prefix to the URL, such as `https://`, `tcp://`, and so forth, if a prefix is not present.
<5> Change the pipelines:
* `spec.pipelines.inputSource` to `spec.pipelines.inputRefs`
* `logs.infra` to `infrastructure`
* `logs.app` to `application`
* `logs.audit` to `audit`
<6> Add a `default` pipeline to send logs to the internal Elasticsearch instance.
+
[NOTE]
====
If you want to forward logs to only the internal {prduct-title} Elasticsearch instance, do not configure the Log Forwarding API.
====

> 

Reviewer: oc create -f ClusterLogForwarder.yaml

.Procedure 

If you use the Log Forwarding API to forward logs, after the Elasticsearch Operator and Cluster Logging Operator are fully updated to 4.6, you must make the following modifications:

. Edit the Cluster Logging custom resource (CR) to remove the `logforwardingtechpreview` annotation:
+
.Sample Cluster Logging CR
[source,yaml]
----
apiVersion: "logging.openshift.io/v1"
kind: "ClusterLogging"
metadata:
  annotations:
    clusterlogging.openshift.io/logforwardingtechpreview: enabled <1>
  name: "instance"
  namespace: "openshift-logging"
....

----
<1> Remove the `logforwardingtechpreview` annotation.

. Export the Log Forwarding CR to create a YAML file for the `ClusterLogForwarder` instance:
+
[source,terminal]
----
$ oc get LogForwarding instance -n openshift-logging -o yaml| tee ClusterLogForwarder.yaml
----

. Edit the YAML file to make the following modifications:
+

> .Sample Log Forwarding CR in {product-title} 4.5

Reviewer 2: I was momentarily confused at the back-to-back code blocks. I suggest swapping the order of the YAML examples and say something like "The modified YAML file resembles the following example:" before this one.

[source,yaml]
----
apiVersion: logging.openshift.io/v1 <1>
kind: ClusterLogForwarder <2>
metadata:
  name: instance
  namespace: openshift-logging
....
spec: <3>
  outputs:
  - url: http://remote.elasticsearch.com:9200 <4>
    name: elasticsearch
    type: elasticsearch
  - url: tls://fluentdserver.example.com:24224
    name: fluentd
    type: fluentdForward <5>
    secret:
      name: fluentdserver
  pipelines:
  - inputRefs: <6>
      - infrastructure
      - application   
    name: mylogs
    outputRefs:
     - elasticsearch
  - inputRefs:
      - audit
    name: auditlogs
    outputRefs:
      - fluentd
      - default <7>
...
----
<1> Change the `apiVersion` from `"logging.openshift.io/v1alpha1"` to `"logging.openshift.io/v1"`.
<2> Change the object kind from `kind: "LogForwarding"` to `kind: "ClusterLogForwarder"`.
<3> Remove the `disableDefaultForwarding: true` parameter.
<4> Change the output parameter from `spec.outputs.endpoint` to `spec.outputs.url`. Add a prefix to the URL, such as `https://`, `tcp://`, and so forth, if a prefix is not present.
<5> For Fluentd outputs, change the `type` from `forward` to `fluentdForward`.
<6> Change the pipelines:

> * `spec.pipelines.inputSource` to `spec.pipelines.inputRefs`

Reviewer 2: I'd add "Change" before each list item.

* `logs.infra` to `infrastructure`
* `logs.app` to `application`
* `logs.audit` to `audit`

> <7> Optional: Add a `default` pipeline to send logs to the internal Elasticsearch instance. You do not need to configure a `default` output.

Reviewer 2: ```suggestion
<7> Optional: Add a `default` pipeline to send logs to the internal Elasticsearch instance. You are not required to configure a `default` output.
```

If you use the Log Forwarding API to forward logs, after the Elasticsearch Operator and Cluster Logging Operator are fully updated to 4.6, you must make the following modifications:

. Edit the Cluster Logging custom resource (CR) to remove the `logforwardingtechpreview` annotation:
+
.Sample Cluster Logging CR
[sourcce,yaml]
----
apiVersion: "logging.openshift.io/v1"
kind: "ClusterLogging"
metadata:
  annotations:
    clusterlogging.openshift.io/logforwardingtechpreview: enabled <1>
  name: "instance"
  namespace: "openshift-logging"
....

----
<1> Remove the remove the `logforwardingtechpreview` annotation.

. Edit the Log Forwarding CR to use some renamed parameters:
+
.Sample Log Forwarding CR in {product-title} 4.5
[sourcce,yaml]
----
apiVersion: "logging.openshift.io/v1alpha1"
kind: "LogForwarding"
metadata:
  name: instance
  namespace: openshift-logging
spec:
  disableDefaultForwarding: true 
  outputs:
   - name: elasticsearch
     type: "elasticsearch"

>      endpoint: https://elasticsearch.openshift-logging.svc:9200 <3>

Reviewer 3: I would suggest using an alternate ns to make it clear this change is not required for the default log store: `https://elasticsearch.production.svc:9200`

  pipelines:
   - name: mylogs
      inputSource:
        - logs.infra
        - logs.app   
        - logs.audit
     outputRefs:
     - elasticsearch
----
+

> .Sample updated Log Forwarding CR in {product-title} 4.6

Author: @sichvoge ^^

[sourcce,yaml]
----
apiVersion: "logging.openshift.io/v1" <1>
kind: "ClusterLogForwarder" <2>
metadata:
  name: instance
  namespace: openshift-logging
spec: <3>
  outputs:
   - name: elasticsearch
     type: "elasticsearch"

>      url: https://elasticsearch.openshift-logging.svc:9200 <4>

Reviewer 3: equivalent here

  pipelines:
   - name: mylogs
      inputRefs: <5>
        - infrastructure
        - application   
        - audit

>      outputRefs:

Reviewer 3: @alanconway thoughts on referencing the default store here?  This sample would otherwise stop routing messages to the default correct?
